### DISCUSSION TITLE
What is the "best practice" to disable form fields while it's submitting?

### DISCUSSION THREAD

**Discussion by catamphetamine** [react-hook-form/react-hook-form#3024] [Q&A] [answered]
There seems to be no such section in the "Advanced" documentation on the website.
If one simply sets `disabled={disabled || formState.isSubmitting}` then the "focusing errored field" feature wouldn't work due to the field being `disabled` (as `disabled` fields are non-focusable) which is not acceptable.
So is there an officially approved way of implementing such a feature?
The one that doesn't involve wrapping everything in `<Controller/>`s:

```js
<Controller 
  onFocus={() => {
    inputRef.current.disabled = false;
    inputRef.current?.focus();
  }}
/>
```

**Comment by bluebill1049** [comment]
can you provide some more detail with a codesandbox as well?

**Comment by catamphetamine** [comment, marked as answer]
There's a form with several fields and when a user clicks "Submit" they can still edit the form fields which is a subpar "user experience". Also, the "Submit" button itself is clickable while the form is submitting, which is also a subpar "user experience" (and a possible source of bugs).

One could add `disabled={formState.isSubmitting}` to every input field and submit button, but that would make this library lose focus after the form has been submitted because `disabled` fields aren't focusable and it calls `.focus()` when submit finishes.

The solution would be either this library delaying `.focus()` until all fields are un-disabled, or manually setting their `disabled` attribute to `false` before calling `.focus()` on a field or submit button.

**Reply by bluebill1049** [reply]
set `shouldFocusError: false` and manage your own custom focus event is the better option here.

**Reply by catamphetamine** [reply]
I see, so no built-in solution for disabling a form while it's submitting. This issue can be closed then.

**Reply by bluebill1049** [reply]
yes, there is no building solution for disabling the button rather than those `formState`.

**Reply by ulken** [reply]
@catamphetamine Wrap all the form fields in a `<fieldset>` (i.e. the only first-level child of the form) which you disable while submitting. The native browser behavior is to disable all its nested fields.

```jsx
const [handleSubmit, formState: { isSubmitting } } = useForm();
const onSubmit = (data) => {         
  // ... submit data
};
const onError = () => {
  // focus field, potentially in a `setTimeout`
};

<form onSubmit={handleSubmit(onSubmit, onError)}
  <fieldset disabled={submitting}>
    <input />
    <select>...</select>
    <button type=”submit”>Submit</button>
  </fieldset>
</form>
```

**Reply by catamphetamine** [reply]
Disabling fields won't work because a disabled field is non-focusable.
Read the original issue description.

**Reply by ulken** [reply]
Sure, but you should be able to use the `onError` handler to manually focus the field, no? Updated the example code.

**Reply by catamphetamine** [reply]
> Sure, but you should be able to use the onError handler to manually focus the field, no?

Your example is just hypothetical and there's no actual code, and you don't know if it works, and you haven't tested it.
I could also propose my own solution using `readOnly` hacking instead of `disabled`, or implementing every field through a `Controller`.
What I was asking (again, you can read the original post) is whether there's an officially approved way of dealing with this bug and whether there's an officially supported approved tested working example code. Which there's none.

**Reply by bluebill1049** [reply]
I don't think we should have this kind of negative attitude towards the user (@ulken) who tries to help and provide a solution. Even it may not be the correct solution, we should still be appreciative and kind, he spends his time trying to help. Can you imagine if we all been negative about people's response who is going to help others?

**Back to the topic**

I think we can extend Controller and subscribe to `formState.isSubmitting` and disable those inputs. 

here is an example of how to extend Controller: https://react-hook-form.com/advanced-usage#ExtendController

**Reply by catamphetamine** [reply]
@bluebill1049 Oh where's the "negative attitude towards the user (@ulken)" then? Are you trying to police how people communicate in this repo? Like a nazi? Ok then, don't take me to the concentration camp, mr. officer.

**Reply by bluebill1049** [reply]
1. I am not policing 
2. Not a nazi
3. There is no concentration camp
4. Not an officer.
5. you can get out of here with your attitude.

**Reply by bluebill1049** [reply]
https://github.com/react-hook-form/react-hook-form/blob/master/CODE_OF_CONDUCT.md
